File: preprocess.py

# preprocess.py
# Example preprocessing for PPG signals.

import numpy as np
from scipy.signal import butter, filtfilt

def butter_bandpass(lowcut, highcut, fs, order=4):
    nyq = 0.5 * fs
    low, high = lowcut / nyq, highcut / nyq
    b, a = butter(order, [low, high], btype='band')
    return b, a

def bandpass_filter(data, lowcut=0.5, highcut=5.0, fs=64, order=4):
    b, a = butter_bandpass(lowcut, highcut, fs, order=order)
    return filtfilt(b, a, data)

def normalize(signal):
    return (signal - np.mean(signal)) / (np.std(signal) + 1e-8)

File: preprocess.m

% preprocess.m
% Preprocessing functions for PPG emotion recognition
% Includes: Bandpass filter + Normalization

function [filteredSignal, normSignal] = preprocess(signal, fs)
    % INPUT:
    %   signal : raw PPG signal (vector)
    %   fs     : sampling frequency (Hz)
    % OUTPUT:
    %   filteredSignal : bandpass filtered PPG
    %   normSignal     : normalized (zero-mean, unit variance)

    % ---- Bandpass filter (0.5 - 5 Hz) ----
    lowcut = 0.5;  % Hz
    highcut = 5.0; % Hz
    order = 4;

    [b, a] = butter(order, [lowcut, highcut]/(fs/2), 'bandpass');
    filteredSignal = filtfilt(b, a, double(signal));

    % ---- Normalization ----
    mu = mean(filteredSignal);
    sigma = std(filteredSignal) + 1e-8; % avoid div-by-zero
    normSignal = (filteredSignal - mu) / sigma;
end
